### ISSUE TITLE
 TCP-over-IPv4

### ISSUE THREAD

**Issue by jcampbell05** [etingof/pysnmp#56]
I am working on a company which uses pysnmp for a cloud printing solution - aiwip.com

We are finding that UCP is causing false negatives in our production codebase since we treat not being able to query the printer as an error.

I read here that TCP support is something of a future development http://pysnmp.sourceforge.net/development.html

What would be the steps needed for us to contribute this to the project to add support for TCP ?

**Comment by etingof** [comment]
Thank you for using pysnmp!

Could you elaborate a bit on your use-case? Is it that you are trying to monitor distant printers which time out occasionally? Why can't you retry?

While TCP might be more reliable, do the printers you monitor support SNMP-over-TCP?

Could you potentially reposition your SNMP managers closer to the printers to reduce packet loss?

**Comment by jcampbell05** [comment]
We are using it to monitor distant printers, for us having the code to manage our printers on the server simplifies our setup. We used to have it on embedded devices connected to the printer but that added overhead of having to remotely update them especially when the network can be down.

Right now we do retry up to 3 times but on some of our 3G based networks it seems the packet drop can be just long enough to thwart all 3 times. I guess we could bump up the amount of retries up to something like 10. The network is connected though since our TCP based connections get through.

I'll look into the SNMP-Over-TCP support, I just assumed all of them supported it these days.

**Comment by etingof** [comment]
> We are using it to monitor distant printers, for us having the code to manage our printers
>  on the server simplifies our setup.

Indeed! Let me rephrase my question -- could you have servers distributed around the world so that servers reside closer to the printers they manage?

In such case I'd suggest you taking a look at the [distributed SNMP proxy](https://github.com/etingof/snmpfwd) tool -- it lets you turn UDP into TCP (for the unreliable part of the network) and then back into UDP for the last mile.

> I'll look into the SNMP-Over-TCP support, I just assumed all of them supported it these days.

No! It's largely optional, AFAIK.

Let me know if you figure your printers do support SNMP-over-TCP. I'm definitely interested in having that implemented to pysnmp .

**Comment by lextm** [comment]
Implemented in v7.1.28. Will ship in a few weeks. This can be closed now.
